File: cpp/Problem1302.cpp
/*
 * Problem: 1302. Deepest Leaves Sum
 * URL: https://leetcode.com/problems/deepest-leaves-sum/
 * Difficulty: Medium
 * Tags: Tree, DFS
 */

#include <vector>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode() : val(0), left(nullptr), right(nullptr) {}
	TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
	TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
	int deepestLeavesSum(TreeNode* root) {
		vector<TreeNode*> nextLevel = {root};
		vector<TreeNode*> currentLevel;
		while (nextLevel.size() != 0) {
			currentLevel = nextLevel;
			nextLevel.clear();
			for (TreeNode* node : currentLevel) {
				if (node->left != nullptr) {
					nextLevel.push_back(node->left);
				}
				if (node->right != nullptr) {
					nextLevel.push_back(node->right);
				}
			}
		}
		int sum = 0;
		for (TreeNode* node : currentLevel) {
			sum += node->val;
		}
		return sum;
	}
};

File: python/Problem1302.py
# Problem: 1302. Deepest Leaves Sum
# URL: https://leetcode.com/problems/deepest-leaves-sum/
# Difficulty: Medium
# Tags: Tree, DFS

from typing import List
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right

class Solution:
  def deepestLeavesSum(self, root: Optional[TreeNode]) -> int:
    maxDepth: int = 0
    result: int = 0
    def dfs(node: Optional[TreeNode], depth: int):
      if (node == None):
        return
      dfs(node.left, depth + 1)
      dfs(node.right, depth + 1)
      nonlocal maxDepth, result
      if depth > maxDepth:
        maxDepth = depth
        result = 0
      if depth == maxDepth:
        result += node.val
    dfs(root, 0)
    return result